File: insurance/src/main/java/com/example/entity/Insurance.java
package com.example.entity;

import com.fasterxml.jackson.annotation.JsonInclude;
import io.quarkus.hibernate.orm.panache.PanacheEntityBase;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import java.time.LocalDate;

@Entity
@JsonInclude
public class Insurance extends PanacheEntityBase {
    @Id
    @GeneratedValue
    public Long id;
    public Long insuredPerson;
    public String policyType;
    public LocalDate validFrom;
    public LocalDate validUntil;
    public String vehicleRegistrationNumber;
    public String vehicleType;
    public Long mileage;
    public String drivingExperience;
    public Boolean vehicleOwnership;
    public String vehicleYear;
    public Integer numberOfSpeedViolations;
    public Integer numberOfDUIViolations;
    public Integer numberOfPastAccidents;

    public void mapInsurance(Insurance insurance) {
        this.id = insurance.id;
        this.policyType = insurance.policyType;
        this.insuredPerson = insurance.insuredPerson;
        this.validFrom = insurance.validFrom;
        this.validUntil = insurance.validUntil;
        this.vehicleRegistrationNumber = insurance.vehicleRegistrationNumber;
        this.vehicleType = insurance.vehicleType;
        this.mileage = insurance.mileage;
        this.drivingExperience = insurance.drivingExperience;
        this.vehicleOwnership = insurance.vehicleOwnership;
        this.vehicleYear = insurance.vehicleYear;
        this.numberOfSpeedViolations = insurance.numberOfSpeedViolations;
        this.numberOfDUIViolations = insurance.numberOfDUIViolations;
        this.numberOfPastAccidents = insurance.numberOfPastAccidents;
    }
}

File: ui/store/Insurance.ts
import { defineStore } from "pinia";
import { useClaimStore } from "./Claim";
import { InsuredPerson } from "./InsuredPerson";

export interface Insurance {
    id?: number,
    insuredPerson?: InsuredPerson,
    policyType?: string,
    validFrom?: Date,
    validUntil?: Date,
    vehicleRegistrationNumber?: number,
    vehicleType?: string,
    mileage?: number,
    drivingExperience?: string,
    vehicleOwnership?: boolean,
    vehicleYear?: string,
    numberOfSpeedViolations?: number,
    numberOfDUIViolations?: number,
    numberOfPastAccidents?: number,
};

interface InsuranceStoreState {
    insurances: Insurance[],
    insurancesSpecific: Insurance[],
    insurance: Insurance
}

export const useInsuranceStore = defineStore("insuranceStore", {
    state : (): InsuranceStoreState => ({
        insurances: [],
        insurancesSpecific: [],
        insurance: {}
    }),
    getters: {
        countActive: (state: InsuranceStoreState) => () => 
            state.insurances.filter(insurance => insurance.validUntil == null).length,
        countCancelled: (state: InsuranceStoreState) => () => 
            state.insurances.filter(insurance => insurance.validUntil != null).length,
    },
    actions: {
        async fetchAll() {
            const config = useRuntimeConfig();
            await fetch(`${config.INSURANCE_BASE_API}`).then(res => res.json()).then(data => this.insurances = data).catch(err => console.log(err));
        },
        async fetchById(id: any) {
            const config = useRuntimeConfig();
            await fetch(`${config.INSURANCE_BASE_API}/${id}`).then(res => res.json()).then(data => this.insurance = data).catch(err => console.log(err));
        },
        async fetchByIdIfProvided(id: any, idProvided: Boolean) {
            this.insurance = {};
            if(idProvided) {
                this.fetchById(id);
            }
        },
        async fetchByInsuredPerson(insuredPersonId: any) {
            const config = useRuntimeConfig();
            await fetch(`${config.INSURANCE_BASE_API}/insuredPerson/${insuredPersonId}`).then(res => res.json()).then(data => this.insurances = data).catch(err => console.log(err));
        },
        async fetchActiveForInsuredPerson(insuredPersonId: any) {
            const config = useRuntimeConfig();
            await fetch(`${config.INSURANCE_BASE_API}/insuredPerson/active/${insuredPersonId}`).then(res => res.json()).then(data => this.insurancesSpecific = data).catch(err => console.log(err));
        },
        async add(insurance: Insurance) {
            const config = useRuntimeConfig();
            await fetch(`${config.INSURANCE_BASE_API}`, {method: "POST", body: JSON.stringify(insurance), headers: {
                'Content-Type': 'application/json;charset=UTF-8'
            }}).then(res => console.log(res)).catch(err => console.log(err));
            this.insurances.push(insurance);
        },
        async update(insurance: Insurance) {
            const config = useRuntimeConfig();
            await fetch(`${config.INSURANCE_BASE_API}/${insurance.id}`, {method: "PUT", body: JSON.stringify(insurance), headers: {
                'Content-Type': 'application/json;charset=UTF-8'
            }}).then(res => console.log(res)).catch(err => console.log(err));
            this.fetchAll();
        },
        async delete(id: any) {
            const config = useRuntimeConfig();

            const claimStore = useClaimStore();
            claimStore.fetchAll();
            claimStore.claims.forEach(claim => {
                if(claim.insurancePolicy == this.insurance.id) {
                    claimStore.delete(claim.id);
                }
            });

            await fetch(`${config.INSURANCE_BASE_API}/${id}`, {method: "DELETE"});
            this.fetchAll();
        },
        async cancel(id: any) {
            const config = useRuntimeConfig();
            await fetch(`${config.INSURANCE_BASE_API}/cancel/${id}`, {method: "PUT"})
            this.fetchById(id);
        }
    }
});



